cmd/juju/crossmodel/list.go

```diff
@@ -103,29 +117,50 @@ func (c *listCommand) Run(ctx *cmd.Context) (err error) {
 	}
 	defer api.Close()
 
-	// TODO (anastasiamac 2015-11-17) add input filters
-	// For now, just filter on the current model.
 	controllerName, err := c.ControllerName()
 	if err != nil {
 		return errors.Trace(err)
 	}
-	modelName, err := c.ClientStore().CurrentModel(controllerName)
+
+	modelName, err := c.ModelName()
 	if err != nil {
-		if errors.IsNotFound(err) {
-			return errors.New("no current model, use juju switch to select a model on which to operate")
-		} else {
-			return errors.Annotate(err, "cannot load current model")
+		return errors.Trace(err)
+	}
+	if !jujuclient.IsQualifiedModelName(modelName) {
```

Comment:
I've refactored the code and also fixed a few other places like model-show and destroy-model to make those use the shared code as well.